util/testutil/integration/dockerd.go

```diff
@@ -0,0 +1,107 @@
+package integration
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/docker/docker/client"
+)
+
+const dockerdBinary = "dockerd"
+
+// InitDockerdWorker registers a dockerd worker with the global registry.
+func InitDockerdWorker() {
+	Register(&dockerd{})
+}
+
+type dockerd struct{}
+
+func (c dockerd) Name() string {
+	return dockerdBinary
+}
+
+func (c dockerd) New(cfg *BackendConfig) (b Backend, cl func() error, err error) {
+	if err := lookupBinary(dockerdBinary); err != nil {
+		return nil, nil, err
+	}
+	if err := lookupBinary("buildkitd"); err != nil {
+		return nil, nil, err
+	}
+	if err := requireRoot(); err != nil {
+		return nil, nil, err
+	}
+
+	deferF := &multiCloser{}
+	cl = deferF.F()
+
+	defer func() {
+		if err != nil {
+			deferF.F()()
+			cl = nil
+		}
+	}()
+
+	tmpdir, err := ioutil.TempDir("", "bktest_dockerd")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	deferF.append(func() error { return os.RemoveAll(tmpdir) })
+
+	configFile := filepath.Join(tmpdir, "config.toml")
+	address := filepath.Join(tmpdir, "dockerd.sock")
+	cfgFile, err := os.Create(configFile)
+	if err != nil {
+		return nil, nil, err
+	}
+	err = json.NewEncoder(cfgFile).Encode(struct {
+		Debug  bool     `json:"debug"`
+		Socket string   `json:"socket"`
+		Hosts  []string `json:"hosts"`
+	}{
+		Debug:  true,
+		Socket: address,
+		Hosts:  []string{address},
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cmd := exec.Command(dockerdBinary, "--config", configFile)
+	logs := map[string]*bytes.Buffer{}
+	ctdStop, err := startCmd(cmd, logs)
+	if err != nil {
+		return nil, nil, err
+	}
+	if err := waitUnix(address, 5*time.Second); err != nil {
+		ctdStop()
+		return nil, nil, fmt.Errorf("dockerd did not start up: %q, %s", err, formatLogs(logs))
+	}
+	deferF.append(ctdStop)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	deferF.append(func() error { cancel(); return nil })
+
+	dockerAPI, err := client.NewClientWithOpts(client.WithHost(address))
+	if err != nil {
+		return nil, nil, err
+	}
+	deferF.append(dockerAPI.Close)
+
+	conn, err := dockerAPI.DialHijack(ctx, "/grpc", "h2c", nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return backend{
+		address:  conn.LocalAddr().String(),
```

Comment:
This can't work like this. The hijacked connection will never have an address where another client can dial into. You would need to create a blank server on a temp socket address that you return. And on every connection to that server call `dockerAPI.DialHijack` and proxy the bytes into the hijacked connection.

Reply:
I started working on this this morning. It's a bit jank, and needs some error handling but I was just playing around with it. Marking this as resolved for now.